lib/sensors/mux.hpp

```diff
@@ -29,10 +29,10 @@ class Mux {
   core::Result selectChannel(const std::uint8_t channel);
   core::Result closeAllChannels();
   core::ILogger &log_;
-  io::I2c i2c_;
+  io::I2c &i2c_;
   const std::uint8_t max_num_unusable_sensors_;
   const std::uint8_t mux_address_;
-  const std::array<std::unique_ptr<II2cMuxSensor<T>>, N> sensors_;
+  const std::array<std::unique_ptr<II2cMuxSensor<T>>, N> &sensors_;
```

Comment:
I feel like this is dangerous. Surely we are fine with the mux owning the sensors? You don't want to be required to make sure that the sensors aren't deallocated while the mux is still around.

Moving the unique pointers into the class should be sufficient.

Reply:
Thing is, if we don't do this, the unique pointers get deleted as soon as we do `sensors_(std::move(sensors))`. This error doesn't come up when we do templating which is what is concerning.

Reply:
What do you mean they get deleted? They're just moved into the array. Of course you can't access the old pointers anymore.

Reply:
Essentially [this](https://stackoverflow.com/questions/39724272/error-call-to-implicitly-deleted-copy-constructor-of-std-1unique-ptra-s#:~:text=Your%20constructor%20argument%20is%20pass%20by%20value%20which%20will%20make%20a%20copy%2C%20but%20you%20cannot%20copy%20a%20std%3A%3Aunique_ptr.%20Passing%20by%20reference%20should%20work%3A)

Reply:
Now, if we do this and try to do `sensor  = sensors_.at(i)` without lhs referencing (this is without templating)
![image](https://user-images.githubusercontent.com/58626285/203565879-d4e606e5-65a2-42a8-a678-d217d1011226.png)

Reply:
There is a fix in #58. It required some housekeeping to make a test that actually instantiates the class. One thing you've done wrong, which I have failed to notices, is that you've put your templates in .cpp files. This is not what you want! A .cpp only containing templates is useless as there will be no code generated and the templates aren't accessible to other compilation units.

Reply:
The TL;DR of the fix is: Don't make your sensors array `const` if you're going to move them.

Reply:
Okay keeping everything to `.hpp` is something I completely didn't realise you could do with templating but makes a lot of sense. Also agree with storing `sensors_` and not a reference without `const` (did not get why it kept copying but this makes sense). Thanks for fixing tho.